src/GatewayAPI/Database/TransactionQuerier.cs

```diff
@@ -332,37 +344,113 @@ private async Task<List<long>> GetAccountTransactionStateVersions(AccountTransac
 
     private async Task<List<Gateway.TransactionInfo>> GetTransactions(List<long> transactionStateVersions)
     {
-        var transactionWithOperationGroups = await _dbContext.LedgerTransactions
-            .Where(t => transactionStateVersions.Contains(t.ResultantStateVersion))
-            .Include(t => t.SubstantiveOperationGroups)
-            .ThenInclude(op => op.InferredAction!.Resource)
-            .Include(t => t.SubstantiveOperationGroups) // https://stackoverflow.com/a/50898208
-            .ThenInclude(op => op.InferredAction!.Validator)
-            .Include(t => t.SubstantiveOperationGroups) // https://stackoverflow.com/a/50898208
-            .ThenInclude(op => op.InferredAction!.FromAccount)
-            .Include(t => t.SubstantiveOperationGroups) // https://stackoverflow.com/a/50898208
-            .ThenInclude(op => op.InferredAction!.ToAccount)
-            .Include(t => t.RawTransaction)
-            .OrderByDescending(lt => lt.ResultantStateVersion)
-            .AsSplitQuery() // See https://docs.microsoft.com/en-us/ef/core/querying/single-split-queries
-            .ToListAsync();
+        async Task<List<LedgerTransaction>> LoadLedgerTransactions(List<long> ids)
+        {
+            await using var localCtx = await _dbContextFactory.CreateDbContextAsync();
+
+            // we're using explicit transaction here to ensure that:
+            // a) SET LOCAL call does not leak outside of this transaction
+            // b) EF will actually reuse same database connection for both queries
+            await using var tx = await localCtx.Database.BeginTransactionAsync();
+
+            // query below tend to operate on quite significant volume of data (over 4 MiB)
+            // so in order to prevent HDD/SSD usage while applying ORDER BY clause
+            await localCtx.Database.ExecuteSqlRawAsync("SET LOCAL work_mem = '32MB';");
+
+            var result = await localCtx.LedgerTransactions
+                .Where(t => ids.Contains(t.ResultantStateVersion))
+                .Include(t => t.SubstantiveOperationGroups)
+                .OrderByDescending(lt => lt.ResultantStateVersion)
+                .ToListAsync();
+
+            await tx.CommitAsync();
+
+            return result;
```

Comment:
Can we add an assert that we get the correct number of transactions back? (Or in particular, if you map to state version and then take a set/distinct, we get a matching number)

This guarantees that this DbContext is up-to-the-same-place as the query-global DbContext, and so we are returning an atomic query at the quoted state version.

Reply:
Added